ops/model.py

```diff
@@ -236,6 +236,10 @@ def __init__(self, relation_name, relation_id, our_unit, backend, cache):
         except RelationNotFoundError:
             # If the relation is dead, just treat it as if it has no remote units.
             pass
+        # If an application has not been determined based on a remote unit, the only remaining
+        # case is a single unit and its peer relation where the application is our unit's application.
+        if self.app is None:
+            self.app = our_unit.app
```

Comment:
On the Juju side, we've been talking about having a relation-created/relation-joined with the app tag, or some other indication to the unit that there *is* a relation, just no remote units have been started.
So while this is probably true in your peer relation case, I don't know if we can trust it to always be true. I realize that we don't yet do the work on the Juju side, but it seems fairly likely to end up in the 2.8 series. (we want to have a nice way to not get Blocked when you have a charm that is related, but the remote units haven't come up yet.)
Thoughts?

One option would be to look for JUJU_APPLICATION_NAME, another option would be to look for metadata relation.peer

Reply:
I agree that this is likely to break for non-peer relations when we do get the `relation-created` hook.  However, I don't think we can use `JUJU_APPLICATION_NAME` or any other relation environment variable, because these lists will also be populated outside of relation hook contexts.

Some other options:

* Check the relation in question against the metadata to see if it's a peer relation (probably hard to reach the meta from this location without some refactoring)
* Get the application name during the relation hook via an env var and then persist it ourselves
* Depend on Juju to provide the application name consistently from something other than (or as an additional field on) `relation-list`

Reply:
I agree with @johnsca regarding `JUJU_APPLICATION_NAME` as I encountered this exactly in the case described when the event wasn't a relation event.

The use-case is a replacement of leader data when you put something into the peer app relation data bag in a non-relation event.

I also considered looking at a list of peer relations via charm metadata but, yes, this will require more significant changes.

It seems right to go down the route of inspecting charm metadata because we have access to it locally and it is deterministic. Quite possibly `is_peer` could be a variable passed into `__init__` of a `Relation` object as an alternative to passing metadata of the whole charm.

@jameinel @johnsca I could suggest a change with `is_peer` if you agree in general.

Reply:
Modified this based on relation roles: https://github.com/canonical/operator/pull/107/commits/da505df72b044a39f059c5444337b27b1f4f2be8

We could also change role names to "provider", "requirer", "peer" similar to how this is done in `juju/charm` https://github.com/juju/charm/blob/b0d0ce63cbe0c6e31d38e886a8c7cb20979b509d/meta.go#L42-L46